max30102.c

```diff
@@ -0,0 +1,159 @@
+                               data) != MAX30102_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t MAX30102_WriteRegister(I2C_HandleTypeDef *hi2c, uint8_t reg, uint8_t value)
+{
+    //write a single byte to the specified register
+    if (HAL_I2C_Mem_Write(hi2c,
+                          MAX30102_I2C_ADDRESS,
+                          reg,
+                          I2C_MEMADD_SIZE_8BIT,
+                          &value,
+                          1,
+                          HAL_MAX_DELAY) != HAL_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t MAX30102_ReadRegister(I2C_HandleTypeDef *hi2c, uint8_t reg, uint8_t *value)
+{
+    //read a single byte from the specified register
+    if (HAL_I2C_Mem_Read(hi2c,
+                         MAX30102_I2C_ADDRESS,
+                         reg,
+                         I2C_MEMADD_SIZE_8BIT,
+                         value,
+                         1,
+                         HAL_MAX_DELAY) != HAL_OK)
+    {
+        return MAX30102_ERROR;
+    }
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t MAX30102_ReadFIFO(I2C_HandleTypeDef *hi2c, uint32_t *red_data, uint32_t *ir_data) {
+    uint8_t fifo_data[6]; 
+
+    // Read 6 bytes from the FIFO data register
+    if (HAL_I2C_Mem_Read(hi2c,
+                         MAX30102_I2C_ADDRESS,
+                         MAX30102_REG_FIFO_DATA,
+                         I2C_MEMADD_SIZE_8BIT,
+                         fifo_data,
+                         6,
+                         HAL_MAX_DELAY) != HAL_OK)
+    {
+        return MAX30102_ERROR;
+    }
+    
+    // Intensity of red LED reflected back into MAX30102 photodetector
+    *red_data = ((uint32_t)fifo_data[0] << 16) | ((uint32_t)fifo_data[1] << 8) | fifo_data[2];
+    // Intensity of infrared LED reflected back into MAX30102 photodetector
+    *ir_data = ((uint32_t)fifo_data[3] << 16) | ((uint32_t)fifo_data[4] << 8) | fifo_data[5]; 
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t ON(I2C_HandleTypeDef *hi2c) {
+    uint8_t data;
+
+    // Read current value of the SHDN bit
+    if (MAX30102_ReadRegister(hi2c, MAX30102_REG_MODE_CONFIG, &data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+    data &= ~0x80; // Wake up sensor by clearing SHDN bit
+    if (MAX30102_WriteRegister(hi2c, MAX30102_REG_MODE_CONFIG, data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+
+
+    return MAX30102_OK;
+}
+
+MAX30102_Status_t OFF(I2C_HandleTypeDef *hi2c) {
+    uint8_t data;
+
+    // Read current value of the SHDN bit
+    if (MAX30102_ReadRegister(hi2c, MAX30102_REG_MODE_CONFIG, &data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+    data |= 0x80; // Put sensor to sleep using SHDN bit
+    if (MAX30102_WriteRegister(hi2c, MAX30102_REG_MODE_CONFIG, data) != MAX30102_OK) {
+        return MAX30102_ERROR;
+    }
+
+
+
+    return MAX30102_OK;
+}
```

Comment:
> Note: The highest bit for registers in a MAX30102 sensor is bit 7 while the lowest is bit 0. Registers are all 8 bits.

ON function relies on the SHDN bit (bit 7 of Mode configuration) value status to be checked, and set to wake up by being cleared. The `ReadRegister` function is responsible for checking the current bit value while the `WriteRegister` clears SHDN. To wake up, `data &= ~0x80` sets bit 7 to 0 and all other bits to 1 (0111 1111)

OFF function works similarly, the only difference is the sensor is put to sleep using the SHDN bit by setting `data |= 0x80` which sets bit 7 to 1 (1000 0000). When the shutdown bit is 1, low-power mode/sleep is enabled.